Nimble Rich Text Editor and Viewer Spec Document (#1289)
ni/nimble: packages/nimble-components/src/rich-text-editor-viewer/specs/README.md

# Nimble Rich Text Editor and Nimble Rich Text Viewer

## Overview

-   The `nimble-rich-text-editor` component is a text editing tool that offers a comprehensive range of functionalities for
    incorporating rich text content. It provides users with the ability to apply different text styles and formats to their content.
-   The `nimble-rich-text-viewer` component is way to view all the rich text content.

### Background

> 

Reviewer: We can write a brief two liner on the potential use cases that has driven the idea of building these components. The background or need for these components are obviously driven by
1. Comments
2. Description

Get some inspirations from [Comments Epic Description](https://ni.visualstudio.com/DevCentral/_backlogs/backlog/ASW%20SystemLink%20Platform/Initiatives/?workitem=2205215)

Author: Added a brief point to elaborate on the background for this component addition.

[Nimble issue #1288](https://github.com/ni/nimble/issues/1288)

[Comments UI mockup](https://www.figma.com/file/Q5SU1OwrnD08keon3zObRX/SystemLink?type=design&node-id=6280-94045)

### Non-goals

> -   The visual and interaction design for `nimble-rich-text-editor` and `nimble-rich-text-viewer` is out of scope for this spec component.

Reviewer: Let's elaborate this more specifically. 
1. ViD and IxD are out of scope for these components in nimble design system  
2. However we have an IxD workflow for SLE application (the figma mockup you linked above) that we will rely on.

    of the rich-text-editor with Angular applications, achieved through a directive, will be developed following the completion of this
    component.

### Features

The `nimble-rich-text-editor` supports the basic text formatting options that includes:

1. Bold
2. Italics
3. Strikeout
4. Numbered List
5. Bulleted List

This component will also offer APIs and interactive methods to format texts in the following ways:

1. Using the buttons located below the editor to modify the text formatting for the selected texts.
2. Using the keyboard shortcuts to modify the text formatting in the editor.
3. Support to get markdown output from the editor.
4. Support to convert the markdown to HTML to render it back in the `nimble-rich-text-viewer` component.

#### _Additional features out of scope of this spec_

-   Allowing the user to tag or mention by entering `@` in the editor and selecting the user name from the drop-down list.
-   Support for adding images to the editor either by uploading or by pasting it.
-   Support for adding hyperlinks to the existing words in the editor. However, adding raw links by pasting to the editor is supported.

> -   Support for underlining the text.

Reviewer: Since we are putting underline support out of scope which is typically done using `<u>` tag of HTML in any markdown, this eliminates our dependency to use html within markdown for current set of features. 

If there are options in tip tap to switch off HTML tags in markdown, it will be worthy to use it and mention it in security section which will completely disallow injection attacks using `<script>` tags into our component, until we need HTML tags

Author: Deferring this comment as it is captured in the security section of this doc.

> -   Provide APIs to make the buttons in the toolbar located below the editor to be hidden or disable.

Reviewer: ```suggestion
-   Provide APIs to configure visibility/disabled state of toolbar buttons for rich text editor
```

### Risks and Challenges

> -   Visual design and interaction design for both the components are out of scope for the initial release. We will start with the existing
>     [mockup screens](https://www.figma.com/file/Q5SU1OwrnD08keon3zObRX/SystemLink?type=design&node-id=6280-94045) which is designed for
>     comments feature. At this moment, we cannot afford to cover all the corner case of visual design specs but we managed to refer from
>     other visual designs and try to incorporate with the nimble standards.

Reviewer: ```suggestion
-   Visual design and interaction design for both the components are out of scope for the initial release. We will start with the existing
    [mockup screens](https://www.figma.com/file/Q5SU1OwrnD08keon3zObRX/SystemLink?type=design&node-id=6280-94045) which is designed for
    comments feature considering customer requirements and deadlines. At this moment, we cannot afford to cover all the corner cases of visual design specs. However we are organically referring to existing nimble components like `button`, `text-area` to provide a consistent visual design
```

> -   Due to the immediate requirement for GM (`Comments Feature`), we are not adding any additional enhancements or feature requirements
>     apart from the mentioned features in the initial release.

Reviewer: Be careful on mentioning customer names in open source libraries. 

```suggestion
-   Due to immediate requirements for comments feature from a business customer, any additional enhancements or requirements apart from whatever is mentioned in this spec are deferred to future scope. 
```

-   The mobile view of the component has not been designed yet, and we are actively collaborating with the design team to create basic mockup screens. We will

>     update this spec accordingly as progress is made.

Reviewer: ```suggestion
    update this spec accordingly based on the progress.
```

### Prior Art/Examples

Sample Mockup Screenshots

_`nimble-rich-text-editor`_

![Editor](./spec-images/editor-sample.png)

_`nimble-rich-text-viewer`_

> ![Viewer](./spec-images/viewer-sample.png)

Reviewer: Let's crop the top header division with username, date and edit/delete menu as per latest discussion of scoping it into application workflow.

Here is the basic high level template used in this component:

```html
<template>
    <section>
        <div id="editor"></div>
    </section>
    <footer>
        <div id="toolbar">
            <nimble-button></nimble-button>
        </div>
        <div>
            <slot name="footer-actions"></slot>
        </div>
    </footer>
</template>
```

An example usage of the `nimble-rich-text-editor` in the application layer is as follows:

```html
<nimble-rich-text-editor>
    <div slot="footer-actions">
        <nimble-button>Cancel</nimble-button>
        <nimble-button>Add Comment</nimble-button>
    </div>
</nimble-rich-text-editor>
```

_nimble-rich-text-viewer_

The `nimble-rich-text-viewer` is used for viewing rich text content when markdown string is passed to it. It performs the post-processing
tasks to convert the markdown string to corresponding HTML nodes for each text formatting.

### API

#### `nimble-rich-text-editor`

_Props/Attrs_

-   none

_Methods_

-   `setMarkdown()` - sets the markdown input from the consumer component to preload the content in the editor.
-   `getMarkdown()` - gets the processed markdown output from the component.

> -   [`destroy()`](https://tiptap.dev/api/editor#destroy) - stops the editor instance and unbinds all events.

Reviewer: Any need for exposing `destroy`? Do we need to handle it internally on detaching this custom element?

Author: Yes, it can be handled internally inside the custom element by destroying the editor in `disconnectedCallback` lifecycle hooks.

> -   [`isEmpty()`](https://tiptap.dev/api/editor#is-empty) - obtains a boolean value indicating whether the editor is empty or not.

Author: Yes, we can emit the `isEmpty` status through the `onChange` event. I have removed the same in the document.

    1. It is a container that allows a client to easily place buttons at the right bottom of the component to interact with the editor.
    2. If no content is slotted in the footer-actions, the element will be emptied and shrunk to accommodate buttons from the
       toolbar.
    3. If content is slotted in the footer-actions, it will occupy a maximum of thirty percent of the entire horizontal footer. If
       there are additional elements beyond this limit, they will be positioned below within the same footer-actions container.

_Note_: The positioning of these slot elements in the mobile view of the component has not yet been confirmed.

_Host Classes_

-   none

_Slotted Content/Slotted Classes_

-   none

_CSS Parts_

-   none

### Angular integration

An Angular directive will be created for both components. The components will not have form association, so a `ControlValueAccessor` will
not be created.

### Blazor integration

A Blazor wrapper will not be created for the component.

### Visual Appearance

NA

## Implementation

We have chosen to utilize the [`Tiptap`](https://tiptap.dev/) rich text editor as the underlying third-party library for developing the

> `nimble-rich-text-editor`. This decision was made due to its extensive range of customization options over other third party libraries for

Author: We have added the same content here as well for your information.

rich text editing and it is the ideal choice for meeting our specific use cases and some are mentioned below:

1. Includes all [basic functionalities](https://tiptap.dev/) like bold, italics, numbered and bulleted lists etc.
2. Includes support to work seamlessly inside the `shadow root`.
3. [Tiptap](https://tiptap.dev/introduction) is built on top of ProseMirror and so have access to all the powerful APIs provided by
   ProseMirror through [`@tiptap/pm`](https://tiptap.dev/guide/prosemirror).
4. Includes support to render the content of the editor with markdown input.

> 5. Includes support to retrieve the content of the editor as HTML and Markdown output (using

Author: HTML output is not part of the initial scope and is also not in the roadmap. We aligned to have markdown as the only output. If there is any requirements on this we will update the same here.

   [prosemirror-markdown](https://github.com/ProseMirror/prosemirror-markdown)).
6. Includes extensions to support [@mention](https://tiptap.dev/api/nodes/mention) functionality (Future scope).

The `nimble-rich-text-editor` is initialized by creating an instance of the [Editor](https://tiptap.dev/api/editor#introduction) class from
the Tiptap's core library. With that we have access to all the APIs exposed, by utilizing some of their extensions like
[StarterKit](https://tiptap.dev/api/extensions/starter-kit) which is a collection of most popular Tiptap extensions includes bold, italics
and all other basic rich text formatting options. All these options([marks](https://tiptap.dev/api/marks)) can also be accessed individually.

> The rich text content entered in the editor is converted to markdown output using `prosemirror-markdown` serializer. Here is the reference for basic formatting

Author: Updated the doc on the same.

| ----------------------- | -------------------------------------------------------- |
| Tab                     | To reach the editor                                      |
| Shift + Arrow keys      | To select the group of texts in the editor               |
| Ctrl + B                | To make the selected text bold                           |
| Ctrl + I                | To make the selected text italic                         |
| Ctrl + Shift + 7        | To enable the focused paragraph a numbered list          |
| Ctrl + Shift + 8        | To enable the focused paragraph a bulleted list          |
| Tab (For lists)         | To create a sub point within a bulleted or numbered list |
| Shift + Tab (For lists) | To remove a sub point from a bulleted or numbered lists  |

_Keyboard navigation with toolbar buttons focused_

| Key          | Behavior                                                                       |
| ------------ | ------------------------------------------------------------------------------ |
| Space, Enter | Enable the selected text formatting feature and return the focus to the editor |

> | Shift + Tab  | To reach the editor back to focus                                              |

Author: Yes, we will make sure to capture it in the development.

_Note_: All other keyboard interaction determined by the slotted element will not be defined in this document.

### Globalization

Currently, this component does not have any string data to be localized. In the footer menu for text formatting, only icons will be displayed. However,
strings may be used for tooltips to enable localization, which will be managed through the `nimble-tooltip` component.

### Security

> -   none

Reviewer: Consider having the below security items: 
1. Prose mirror uses markdown-it for converting markdown to HTML and HTML to markdown. We will follow the [security guidelines of markdown-it](https://github.com/markdown-it/markdown-it/blob/master/docs/security.md#security) to turn off HTML at source as given in the [API docs](https://markdown-it.github.io/markdown-it/#MarkdownIt.new)
2. For additional safety, use [sanitize-html](https://www.npmjs.com/package/sanitize-html) package to whitelist only specific html tags on need basis or completely disallow any html tags

Author: Added the points as suggested.

### Performance

Performance of tiptap explained in an example of having [long texts](https://tiptap.dev/examples/book).

### Dependencies

This component is dependent on the [`tiptap`](https://tiptap.dev/) third party library. As mentioned,
[tiptap](https://tiptap.dev/introduction#why-should-i-use-tiptap) is built on top of prosemirror, we might also use some of the prosemirror's
library. For the currently supported features, we will include the following libraries that will be added to the packages.json

-   @tiptap/core
-   @tiptap/pm
-   @tiptap/starter-kit

**_Note_**: For markdown parser and serializer, [prosemirror-markdown](https://github.com/ProseMirror/prosemirror-markdown) internal dependencies will be installed along with this.

### Test Plan

The current testing plan involves writing unit tests to verify the core functionality of the component. All tests for this component will
follow the standards set by the nimble repository, which include unit tests to assess component logic and Chromatic tests to cover all
visual states of the component.

### Tooling

> _Are there any special considerations for tooling? Will tooling changes need to be made? Is there a special way to light up this component

Reviewer: Update this section as applicable - NA

    in this spec are deferred to future scope.
-   The mobile view of the component has not been designed yet, and we are actively collaborating with the design team to create basic mockup screens. We will
    update this spec accordingly based on the progress.

### Prior Art/Examples

Sample Mockup Screenshots

_`nimble-rich-text-editor`_

![Editor](./spec-images/editor-sample.png)

_`nimble-rich-text-viewer`_

![Viewer](./spec-images/viewer-sample.png)

## Design

_nimble-rich-text-editor_

The `nimble-rich-text-editor` will be divided into two sections namely an `editor` section and a `footer` section.

1. `editor` section is the actual text area to add or update rich text contents.
2. `footer` section consists of `nimble-button` to control each text formatting functionalities like bold, italic etc,. and a
   `footer-actions` slot element which is typically used to add action buttons to the right bottom of the component.

Here is the basic high level template used in this component:

```html
<template>
    <section>
        <div id="editor"></div>

>     </section>

Author: Updated to section to wrap the entire editor into it.

    <footer>
        <div id="toolbar">
            <nimble-button></nimble-button>
        </div>
        <div>
            <slot name="footer-actions"></slot>

>         </div>

Author: Updated

>         </div>

Author: Yes, we can use section instead of div. I have updated the same.

    </footer>
</template>
```

An example usage of the `nimble-rich-text-editor` in the application layer is as follows:

```html
<nimble-rich-text-editor>
    <div slot="footer-actions">
        <nimble-button>Cancel</nimble-button>
        <nimble-button>Add Comment</nimble-button>
    </div>
</nimble-rich-text-editor>
```

_nimble-rich-text-viewer_

The `nimble-rich-text-viewer` is used for viewing rich text content when markdown string is passed to it. It performs the post-processing
tasks to convert the markdown string to corresponding HTML nodes for each text formatting.

### API

#### `nimble-rich-text-editor`

_Props/Attrs_

-   none

_Methods_

> -   `setMarkdown()` - sets the markdown input from the consumer component to preload the content in the editor.

Reviewer: ```suggestion
-   `setMarkdown()` - sets the markdown input from the consumer component which can be used for preloading the content
```

-   `getMarkdown()` - gets the processed markdown output from the component.
-   [`destroy()`](https://tiptap.dev/api/editor#destroy) - stops the editor instance and unbinds all events.
-   [`isEmpty()`](https://tiptap.dev/api/editor#is-empty) - obtains a boolean value indicating whether the editor is empty or not.

_Events_

-   `onChange` - event emitted when there is a change in the the editor. This can be achieved through tiptap's

>     [transaction event](https://tiptap.dev/api/events#transaction).

Author: For `change` event, we will be using the update event exposed by tiptap. I have updated the same in the doc.

    of the rich-text-editor with Angular applications, achieved through a directive, will be developed following the completion of this
    component.

### Features

The `nimble-rich-text-editor` supports the basic text formatting options that includes:

1. Bold
2. Italics
3. Numbered List
4. Bulleted List

This component will also offer APIs and interactive methods to format texts in the following ways:

1. Using the buttons located below the editor to modify the text formatting for the selected texts.
2. Using the keyboard shortcuts to modify the text formatting in the editor.
3. Support to get markdown output from the editor.
4. Support to convert the markdown to HTML to render it back in the `nimble-rich-text-viewer` component.

#### _Additional features out of scope of this spec_

-   Allowing the user to tag or mention by entering `@` in the editor and selecting the user name from the drop-down list.
-   Support for adding images to the editor either by uploading or by pasting it.
-   Support for adding hyperlinks to the existing words in the editor. However, adding raw links by pasting to the editor is supported.

> -   Support for striking out and underlining text. The reason for including these features in future scope that we utilize the
>     [prosemirror-markdown](https://github.com/ProseMirror/prosemirror-markdown) serializer and parser to convert the text into markdown format and vice
>     versa. However, the supported functionality of prosemirror-markdown, as mentioned in their

Reviewer: ```suggestion
-   Support for striking out and underlining text. We use
    [prosemirror-markdown](https://github.com/ProseMirror/prosemirror-markdown) serializer and parser to convert the text into markdown format and vice
    versa. However, the supported functionality of prosemirror-markdown, as mentioned in their
```

    [documentation](https://github.com/ProseMirror/prosemirror-markdown#documentation), does not include support for strikeout and underline. To

>     address this limitation, we need to extend the class and add the necessary nodes to both the serializer and parser which will be added in the

Reviewer: Is there any tiptap/prose mirror documentation that instructs on applying these extensions for strikeout and underline behaviors? If yes, it will be good to link it here.

    <footer>
        <section id="toolbar">
            <nimble-button></nimble-button>
        </section>
        <section id="actions">
            <slot name="footer-actions"></slot>
        </section>
    </footer>
</template>
```

An example usage of the `nimble-rich-text-editor` in the application layer is as follows:

```html
<nimble-rich-text-editor>
    <div slot="footer-actions">
        <nimble-button>Cancel</nimble-button>
        <nimble-button>Add Comment</nimble-button>
    </div>
</nimble-rich-text-editor>
```

_nimble-rich-text-viewer_

The `nimble-rich-text-viewer` is used for viewing rich text content when markdown string is passed to it. It performs the post-processing
tasks to convert the markdown string to corresponding HTML nodes for each text formatting.

### `nimble-rich-text-editor`

### API

_Props/Attrs_

-   none

_Methods_

-   `setMarkdown()` - sets the markdown input from the consumer component which can be used for preloading the content.
-   `getMarkdown()` - gets the processed markdown output from the component.

_Events_

-   `change` - event emitted when there is a change in the the editor. This can be achieved through tiptap's

>     [update event](https://tiptap.dev/api/events#update). Below is a scenarios of event triggers for the update event, indicating when they do and do not
>     occur:

Reviewer: ```suggestion
    [update event](https://tiptap.dev/api/events#update). Below are few scenarios to understand when an update event will trigger or not trigger:
```

| Space, Enter | Enable the selected text formatting feature and return the focus to the editor |
| Shift + Tab  | To reach the editor back to focus                                              |

_Note_: All other keyboard interaction determined by the slotted element will not be defined in this document.

### Globalization

Currently, this component does not have any string data to be localized. In the footer menu for text formatting, only icons will be displayed. However,
strings may be used for tooltips to enable localization, which will be managed through the `nimble-tooltip` component.

### Security

-   Prose mirror uses markdown-it for converting markdown to HTML and HTML to markdown. We will follow the
    [security guidelines of markdown-it](https://github.com/markdown-it/markdown-it/blob/master/docs/security.md#security) to turn
    off HTML at source as given in the [API docs](https://markdown-it.github.io/markdown-it/#MarkdownIt.new). Prosemirror-markdown follows the same as shown in this
    [specific line of code](https://github.com/ProseMirror/prosemirror-markdown/blob/26e58302399b7d9a9b3bc8fc3bf5807627ca29e5/src/from_markdown.ts#L245).

> -   For additional safety, using [sanitize-html](https://www.npmjs.com/package/sanitize-html) package to whitelist only specific html
>     tags on need basis or completely disallow any html tags.

Reviewer: ```suggestion
-   For additional safety we may use [sanitize-html](https://www.npmjs.com/package/sanitize-html) package on need basis if HTML is turned on. This will whitelist only specific html
    tags needed for rich text markdown. 
```

> # Nimble Rich Text Editor and Nimble Rich Text Viewer

Author: Thanks for the suggestion. We went up with option 2 by having the actual spec in `rich-text-editor` and `rich-text-viewer` spec have the link to the actual spec.

## Overview

-   The `nimble-rich-text-editor` component is a text editing tool that offers a comprehensive range of functionalities for
    incorporating rich text content. It provides users with the ability to apply different text styles and formats to their content.
-   The `nimble-rich-text-viewer` component is way to view all the rich text content.

### Background

Collaboration is a key requirement in order for users to adopt a tool at scale and would help improve overall customer satisfaction with the product.
The Comment Feature, designed for adding and viewing comments with rich text content, should be generic enough to be reused in several workflows.
Therefore, it is essential to develop web components that allow users to view and create rich text content, enabling their use in various scenarios,
including Comments and other instances that necessitate rich text capabilities.

[Nimble issue #1288](https://github.com/ni/nimble/issues/1288)

[Comments UI mockup](https://www.figma.com/file/Q5SU1OwrnD08keon3zObRX/SystemLink?type=design&node-id=6280-94045)

> 

Author: Added AzDo Comments Epic link here.

### Non-goals

-   The visual design and interaction design for these components are out of scope for the initial release. However we have an
    [IxD workflow](https://www.figma.com/file/Q5SU1OwrnD08keon3zObRX/SystemLink?type=design&node-id=6280-94045) for SLE application that we will rely
    on.
-   Blazor integration will be out of scope for this component as we have less or no experience in Blazor component development or other
    related technology. Blazor example app addition will also be not covered in this spec.

> -   Adding an example Angular client application for this component may not be included in the initial releases. However, the integration

Author: Thanks for letting us know. We will add these components to the example client app once the Angular integration is completed. I have now removed this point from the Non-Goals.

    of the rich-text-editor with Angular applications, achieved through a directive, will be developed following the completion of this
    component.

> ### Features

Author: I have added this point in the feature section and linked the markdown flavor we support.

-   Due to immediate requirements for comments feature from a business customer, any additional enhancements or requirements apart from whatever is
    mentioned
    in this spec are deferred to future scope.
-   The mobile view of the component has not been designed yet, and we are actively collaborating with the design team to create basic mockup screens. We
    will update this spec accordingly based on the progress.

### Prior Art/Examples

Sample Mockup Screenshots

_`nimble-rich-text-editor`_

![Editor](./spec-images/editor-sample.png)

_`nimble-rich-text-viewer`_

![Viewer](./spec-images/viewer-sample.png)

## Design

_nimble-rich-text-editor_

The `nimble-rich-text-editor` will be divided into two sections namely an `editor` section and a `footer` section.

1. `editor` section is the actual text area to add or update rich text contents.
2. `footer` section consists of `nimble-button` to control each text formatting functionalities like bold, italic etc,. and a
   `footer-actions` slot element which is typically used to add action buttons to the right bottom of the component.

> Here is the shadow DOM template used in `nimble-rich-text-editor` component:

Author: I have moved this to the Anatomy section.

```html
<template>
    <section id="editor"></section>
    <footer>
        <section id="toolbar">

>             <nimble-button></nimble-button>

Author: Resolving this comment as I have replied to the other comment on the usage of `nimble-toolbar`.

>             <nimble-button></nimble-button>

Author: Added the `nimble-toolbar` for formatting options as suggested.

        </section>
        <section id="actions">
            <slot name="footer-actions"></slot>
        </section>
    </footer>
</template>
```

An example usage of the `nimble-rich-text-editor` in the application layer is as follows:

```html
<nimble-rich-text-editor>
    <div slot="footer-actions">
        <nimble-button>Cancel</nimble-button>
        <nimble-button>Add Comment</nimble-button>
    </div>
</nimble-rich-text-editor>
```

_nimble-rich-text-viewer_

The `nimble-rich-text-viewer` is used for viewing rich text content when markdown string is passed to it. It performs the post-processing
tasks to convert the markdown string to corresponding HTML nodes for each text formatting.

### `nimble-rich-text-editor`

### API

_Props/Attrs_

-   none

_Methods_

-   `setMarkdown()` - sets the markdown input from the consumer component which can be used for preloading the content.
-   `getMarkdown()` - gets the processed markdown output from the component.

_Events_

> -   `change` - event emitted when there is a change in the the editor. This can be achieved through tiptap's

Author: I have updated this event name to `input` which makes more sense compared to the `change` event as it fires for every input when typing. I have updated this accordingly in the spec.

> -   `change` - event emitted when there is a change in the the editor. This can be achieved through tiptap's

Author: @jattasNI and Team, we have added the `isEmpty` attribute as part of the event details to dynamically update the enable state of the "Add comment" button. 

We have referred to the `nimble-banner` [toggle](https://github.com/ni/nimble/blob/rich-text-editor-viewer-spec/packages/nimble-components/src/banner/index.ts#L66) event on how to set the details object in the event that is fired and in which they were using something like `newState`, `oldState` to check whether it is open or not. 

Please let us know your thoughts on this approach.

> -   `change` - event emitted when there is a change in the the editor. This can be achieved through tiptap's

Author: Resolving this comment as we have the same conversation in the other thread.

`prosemirror-markdown` parser, which is then converted to HTML string and rendered into the component to view all rich text content.

### Prototype

[`Tiptap prototype in Stackblitz`](https://stackblitz.com/edit/typescript-g3yfmo?file=index.ts)

The prototype includes the below functionalities,

1. Basic formatting support.
2. Image support. - The prototype includes [`Tiptap Extension`](https://tiptap.dev/api/nodes/image) to support images.
3. `@mention` support. - The prototype includes default user list that gets triggered using `@` character.
4. Markdown support. - The prototype includes support to render/retrieve content in markdown format.
5. `Shadow root` support - The prototype uses [Microsoft Fast](https://github.com/microsoft/fast) to create the rich text editor as a
   custom component that renders in the shadow root.
6. `Top layer` support - Hyperlink using `nimble-dialog` which renders the popup in top layer.

**_Note_**: This prototype is not made in a responsive fashion for smaller screens and is intended only for testing the specific functionalities listed
above.

### States

_Button State_

> The buttons in the toolbar will indicate the formatting status of the selected text in the editor. For example, if the text in the editor

Author: @mollykreis, Initially, when developing the prototype for this component, we tried using the `nimble-toggle-button` but in some cases it does not behave as expected. We wanted the button to toggle the state only based on the condition we set in `checked` Boolean attribute of the component not changing the state by clicking on the button. In one of the scenario, when we were having focus on the sub point of numbered or bulleted list (image 1), clicking the button should not toggle its state as condition attached to `?checked` attribute is still `true`. But it changes its state to `false` (Image 2). We used `nimble-toggle-button` for numbered list in the below snapshots.
![image](https://github.com/ni/nimble/assets/123377523/db5cf2e1-326a-4a49-9432-4f43b96861a6)
Not expected:
![image](https://github.com/ni/nimble/assets/123377523/c01da55e-cfd9-4ce1-abb8-fdd6d463ba4f)

Whereas, when trying with the `nimble-button` by toggling the `appearance` we can achieve it as it does not toggle the appearance on click. We used `nimble-button` for bulleted list in the below snapshots.
![image](https://github.com/ni/nimble/assets/123377523/9be21398-9096-4bb4-b913-74572846fe9e)
Expected:
![image](https://github.com/ni/nimble/assets/123377523/cce89358-e7ce-43cc-b07d-da73b38d561a)

Is this expected for `nimble-toggle-button` to change its state on click and not based on the condition? If yes, can we go with the `nimble-button` approach?

> The buttons in the toolbar will indicate the formatting status of the selected text in the editor. For example, if the text in the editor

Author: Thanks for sharing the `group-row` component as an example for resolving this issue we were able to quickly check it in our end and it worked. So we can go with the `nimble-toggle-button` for the state changes in the toolbar. I have updated the spec accordingly. Adding here the link for the reference:

- template: [nimble/packages/nimble-components/src/table/components/group-row/template.ts at main · ni/nimble · GitHub](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/table/components/group-row/template.ts)
- class: [nimble/packages/nimble-components/src/table/components/group-row/index.ts at main · ni/nimble · GitHub](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/table/components/group-row/index.ts)

All the basic keyboard accessibility for the tiptap rich text editor is mentioned in their
[keyboard shortcut](https://tiptap.dev/api/keyboard-shortcuts#predefined-keyboard-shortcuts) page. To facilitate easy access, listed down
the supported text formatting keyboard shortcuts in the tiptap editor.

| Key                     | Behavior                                                 |
| ----------------------- | -------------------------------------------------------- |
| Tab                     | To reach the editor                                      |
| Shift + Arrow keys      | To select the group of texts in the editor               |
| Ctrl + B                | To make the selected text bold                           |
| Ctrl + I                | To make the selected text italic                         |
| Ctrl + Shift + 7        | To enable the focused paragraph a numbered list          |
| Ctrl + Shift + 8        | To enable the focused paragraph a bulleted list          |
| Tab (For lists)         | To create a sub point within a bulleted or numbered list |
| Shift + Tab (For lists) | To remove a sub point from a bulleted or numbered lists  |

> _Keyboard navigation with toolbar buttons focused_

Reviewer 2: Switching between the buttons in the toolbar is missing here. For example, moving from the "bold" button to the "italic" button. Is that accomplished by pressing "Tab" or by pressing the "Right Arrow" key? I think it should be using the arrow keys, which means if you use the `nimble-toolbar` as I mentioned in another comment, you'll get that behavior for free.

Author: @mollykreis, We have tried using the `nimble-toolbar` in our prototype and here are some of the problems faced in the keyboard accessibility,
1. Consider having the focus in the editor text area and by pressing "Tab" will move the focus to icon buttons in the toolbar. Now by pressing "Arrow keys" to move the focus between formatting buttons. We also have `footer-actions` slot element in the same toolbar at the right end, which we specifically added for the comments feature ([mockup](https://www.figma.com/proto/Q5SU1OwrnD08keon3zObRX/SystemLink?type=design&node-id=6280-95369&scaling=min-zoom&page-id=2428%3A32954&starting-point-node-id=6280%3A94118&show-proto-sidebar=1)) to have `Cancel` and `Add comment` button. To reach the buttons in the slot, we can only use "Tab" because both the buttons are slotted into `slot = footer-actions` and the `footer-actions` is slotted into `slot = end` of nimble-toolbar. Is this making inconsistent in moving the focus by using "Arrow keys" for formatting options and "Tab" for slotted buttons which in the same toolbar?
2. From the comments feature point of view, even after we change the focus from formatting options (icon buttons) to `footer-actions` buttons by pressing "Tab", we couldn't move forward once after that. For an instance, consider the focus in the formatting options, pressing "Tab" moves the focus to `cancel` button and pressing "Tab" again, moves the focus back to the formatting options and not the `Add comment` button. 

Please let us know if there are any alternative design that we should consider to resolve these issue or is it fine to use the section or div tag to achieve the keyboard accessibility through "Tab" in all places.

Author: Thanks for the detailed information, we will use `nimble-toolbar` to wrap all the formatting options and wrap this toolbar and a slot element into another section to maintain it in a flexbox to stick with the mockups. I have updated the same in the [shadow DOM template](https://github.com/ni/nimble/blob/rich-text-editor-viewer-spec/packages/nimble-components/src/rich-text-editor/specs/README.md#anatomy) section of this spec. I have also updated the key accessibility which is to have "arrow keys" for the toolbar and "Tab" to switch to footer-action buttons.

| Key          | Behavior                                                                       |
| ------------ | ------------------------------------------------------------------------------ |
| Space, Enter | Enable the selected text formatting feature and return the focus to the editor |
| Shift + Tab  | To reach the editor back to focus                                              |

_Note_: All other keyboard interaction determined by the slotted element will not be defined in this document.

### Globalization

> Currently, this component does not have any string data to be localized. In the footer menu for text formatting, only icons will be displayed. However,

Reviewer 2: The `nimble-tooltip` is an experimental component, so I wouldn't recommend using it here.

Also, I think it should be a requirement to have tooltips (we've been using the `title` attribute to accomplish this) and accessible text for each icon button. There is a spec in progress for how to configure localizable strings on a component, so once the infrastructure work is complete, I'd expect this component to plug into the shared mechanism for providing localized strings. See #1272

Author: Thanks for the information on localization. I have updated the same in the spec.
